Not possible to sync data with geodiff

I am not able to sync data and it appears to have corrupted my gpkg file:
```
2019-11-24T12:59:33.117Z push saber/CABFAS Saber: Geodiff create changeset on vector/trees.gpkg successful: total size 229 bytes
2019-11-24T12:59:33.120Z push saber/CABFAS Saber: 1 items to upload (total size 229 bytes)
2019-11-24T12:59:33.123Z push saber/CABFAS Saber: Starting push request: https://public.cloudmergin.com/v1/project/push/saber/CABFAS Saber
2019-11-24T12:59:33.239Z push saber/CABFAS Saber: Push request accepted. Transaction ID: 285159f2-0381-4519-b167-63c1e8b06782
2019-11-24T12:59:33.243Z push saber/CABFAS Saber: Uploading item: https://public.cloudmergin.com//v1/project/push/chunk/285159f2-0381-4519-b167-63c1e8b06782/72294e53-86a9-405b-a7eb-0347aeae0283
2019-11-24T12:59:33.328Z push saber/CABFAS Saber: Uploaded successfully: 72294e53-86a9-405b-a7eb-0347aeae0283
2019-11-24T12:59:33.333Z push saber/CABFAS Saber: Requesting transaction finish: 285159f2-0381-4519-b167-63c1e8b06782
2019-11-24T12:59:33.410Z push saber/CABFAS Saber: FAILED - Network API error: uploadFinish(): Error transferring https://public.cloudmergin.com/v1/project/push/finish/285159f2-0381-4519-b167-63c1e8b06782 - server replied: BAD REQUEST. {
    "corrupted_files": [
        "vector/trees.gpkg"
    ]
}

2019-11-24T12:59:33.414Z sync saber/CABFAS Saber: ### FAILED ###

```

Mergin log:
```
[ACCESS] 80.79.16.218, 192.168.101.120 GET /v1/project tags=input_use&flag=created HTTP/1.1 200 8066 - "Mozilla/5.0" 95071 <107>
[ACCESS] 192.168.101.120 HEAD /ping  HTTP/1.1 200 0 - "-" 712 <107>
[ACCESS] 80.79.16.218, 192.168.101.120 GET /v1/project/viktor/multi_dir since=v2 HTTP/1.1 200 1031 - "Mozilla/5.0" 19937 <107>
[ACCESS] 192.168.85.156 HEAD /ping  HTTP/1.1 200 0 - "-" 839 <108>
[ACCESS] 80.79.16.218, 192.168.85.156 POST /v1/project/push/viktor/multi_dir  HTTP/1.1 200 60 - "Mozilla/5.0" 311302 <107>
[ACCESS] 80.79.16.218, 192.168.101.120 POST /v1/project/push/chunk/494d7548-ea61-4e79-9544-8cde2b13b8ff/c7ccd43f-5ebf-4409-8e83-7eacb9942811  HTTP/1.1 200 65 - "Mozilla/5.0" 49909 <108>
[ERROR] [108] Data integrity check has failed on file data/data.gpkg in project viktor/multi_dir||NoneType: None
```
Input log:
```
FAILED - Network API error: uploadFinish(): Error transferring https://dev.dev.cloudmergin.com/v1/project/push/finish/494d7548-ea61-4e79-9544-8cde2b13b8ff - server replied: BAD REQUEST. {\n    \"corrupted_files\": [\n        \"data/data.gpkg\"\n    ]\n}\n\n"
```

--- a/app/merginapi.cpp
+++ b/app/merginapi.cpp
@@ -12,6 +12,7 @@
 
 #include "inpututils.h"
 #include "geodiffutils.h"
+#include "qgsquickutils.h"
 
 #include <geodiff.h>
 
@@ -1602,12 +1603,13 @@ void MerginApi::uploadInfoReplyFinished()
         if ( geodiffRes == GEODIFF_SUCCESS )
         {
           QByteArray checksumDiff = getChecksum( diffPath );
+          //QByteArray checksumBase = getChecksum( basePath );
 
           // TODO: this is ugly. our basefile may not need to have the same checksum as the server's
           // basefile (because each of them have applied the diff independently) so we have to fake it
           QByteArray checksumBase = serverProject.fileInfo( filePath ).checksum.toLatin1();
 
-          merginFile.diffName = QFileInfo( diffPath ).fileName();
+          merginFile.diffName = QgsQuickUtils::getRelativePath( diffPath, transaction.projectDir + "/.mergin/" );
           merginFile.diffChecksum = QString::fromLatin1( checksumDiff.data(), checksumDiff.size() );
           merginFile.diffSize = QFileInfo( diffPath ).size();
           merginFile.chunks = generateChunkIdsForSize( merginFile.diffSize );

--- a/app/merginapi.cpp
+++ b/app/merginapi.cpp
@@ -1603,7 +1603,6 @@ void MerginApi::uploadInfoReplyFinished()
         if ( geodiffRes == GEODIFF_SUCCESS )
         {
           QByteArray checksumDiff = getChecksum( diffPath );
-          //QByteArray checksumBase = getChecksum( basePath );
 
           // TODO: this is ugly. our basefile may not need to have the same checksum as the server's
           // basefile (because each of them have applied the diff independently) so we have to fake it

--- a/app/test/testmerginapi.cpp
+++ b/app/test/testmerginapi.cpp
@@ -944,6 +944,43 @@ void TestMerginApi::testDiffUpload()
   QCOMPARE( MerginApi::localProjectChanges( projectDir ), ProjectDiff() );  // no local changes expected
 }
 
+void TestMerginApi::testDiffSubdirsUpload()
+{
+  QString projectName = "testDiffSubdirsUpload";
+  QString projectDir = mApi->projectsPath() + "/" + projectName;
+
+  createRemoteProject( mApiExtra, mUsername, projectName, mTestDataPath + "/" + "diff_project_subs" + "/" );
+
+  downloadRemoteProject( mApi, mUsername, projectName );
+
+  const QString base( "subdir/subsubdir/base.gpkg" );
+  QVERIFY( QFileInfo::exists( projectDir + "/.mergin/" + base ) );
+
+  QCOMPARE( MerginApi::localProjectChanges( projectDir ), ProjectDiff() );  // no local changes expected
+
+
+  // replace gpkg with a new version with a modified geometry
+  QFile::remove( projectDir + "/" + base );
+  QFile::copy( mTestDataPath + "/modified_1_geom.gpkg", projectDir + "/" + base );
+
+  ProjectDiff diff = MerginApi::localProjectChanges( projectDir );
+  ProjectDiff expectedDiff;
+  expectedDiff.localUpdated = QSet<QString>() << base ;
+  QCOMPARE( diff, expectedDiff );
+
+  GeodiffUtils::ChangesetSummary expectedSummary;
+  expectedSummary["simple"] = GeodiffUtils::TableSummary( 0, 1, 0 );
+  expectedSummary["gpkg_contents"] = GeodiffUtils::TableSummary( 0, 1, 0 );
+  QString changes = GeodiffUtils::diffableFilePendingChanges( projectDir, base, true );
+  GeodiffUtils::ChangesetSummary summary = GeodiffUtils::parseChangesetSummary( changes );
+  QCOMPARE( summary, expectedSummary );
+
+  uploadRemoteProject( mApi, mUsername, projectName );
+
+  QCOMPARE( MerginApi::localProjectChanges( projectDir ), ProjectDiff() );  // no local changes expected
+}
+
+
 void TestMerginApi::testDiffUpdateBasic()
 {
   // test case where there is no local change in a gpkg, it is only modified on the server

--- a/app/test/testmerginapi.h
+++ b/app/test/testmerginapi.h
@@ -44,6 +44,7 @@ class TestMerginApi: public QObject
     void testConflictRemoteAddLocalAdd();
     void testUploadWithUpdate();
     void testDiffUpload();
+    void testDiffSubdirsUpload();
     void testDiffUpdateBasic();
     void testDiffUpdateWithRebase();
     void testDiffUpdateWithRebaseFailed();

